lib/TPP/Transforms/LoopInsertion.cpp

```diff
@@ -0,0 +1,276 @@
+    ReassociationIndices reassociationIndex;
+    for (size_t j = 0; j < tileShapes[i].size(); j++)
+      reassociationIndex.push_back(index++);
+    indices.push_back(reassociationIndex);
+  }
+  for (size_t i = tileShapes.size(); i < origtensorShape.size(); i++) {
+    ReassociationIndices reassociationIndex;
+    reassociationIndex.push_back(index++);
+    indices.push_back(reassociationIndex);
+  }
+
+  return indices;
+}
+
+void insertSubview(ArrayRef<int64_t> tensorShape, Type type, Type resultType,
+                   SmallVector<ReassociationIndices> reassociation,
+                   Value operand, ForallOp op, OpBuilder b,
+                   xsmm::BrgemmOp brgemmOp, int operandNumber) {
+
+  auto expandShape = b.create<memref::ExpandShapeOp>(
+      op.getLoc(),
+      MemRefType::get({tensorShape},
+                      dyn_cast<MemRefType>(type).getElementType()),
+      operand, reassociation);
+  expandShape.setStaticOutputShape(tensorShape);
+  b.setInsertionPoint(brgemmOp);
+  SmallVector<OpFoldResult> strides(tensorShape.size(), b.getIndexAttr(1)),
+      sizes, offsets;
+  size_t tileSize =
+      tensorShape.size() - dyn_cast<ShapedType>(resultType).getShape().size();
+
+  SmallVector<int64_t> tileSizes;
+  for (size_t i = 0; i < tensorShape.size(); i++) {
+    if (i < tileSize) {
+      int opnum = operandNumber;
+      if (opnum == 3) {
+        opnum = 1;
+      }
+      int inductionVarIndex = (opnum - 1) * tileSize + i;
+      offsets.push_back(op.getInductionVars()[inductionVarIndex]);
+      sizes.push_back(b.getIndexAttr(1));
+    } else {
+      sizes.push_back(b.getIndexAttr(tensorShape[i]));
+      tileSizes.push_back(tensorShape[i]);
+      offsets.push_back(b.getIndexAttr(0));
+    }
+  }
+
+  auto subviewType =
+      MemRefType::get({tileSizes}, dyn_cast<MemRefType>(type).getElementType());
+  auto [originalStride, originalOffset] =
+      getStridesAndOffset(dyn_cast<MemRefType>(subviewType));
+  subviewType = MemRefType::get(
+      {tileSizes}, dyn_cast<MemRefType>(subviewType).getElementType(),
+      StridedLayoutAttr::get(b.getContext(), ShapedType::kDynamic,
+                             originalStride));
+  auto subview = b.create<memref::SubViewOp>(
+      op.getLoc(), dyn_cast<MemRefType>(subviewType), expandShape.getResult(),
+      offsets, sizes, strides);
+  brgemmOp.getOperand(operandNumber).replaceAllUsesWith(subview);
+}
+
+FailureOr<ForallOp> insertParallelLoop(ForallOp op,
+                                       ArrayRef<unsigned> tileShapeM,
+                                       ArrayRef<unsigned> tileShapeN) {
+  xsmm::BrgemmOp brgemmOp = NULL;
+  OpBuilder b(op);
+  for (auto oper = op.getBody()->getOperations().begin();
+       oper != op.getBody()->getOperations().end(); oper++)
+    if (dyn_cast<xsmm::BrgemmOp>(oper)) {
+      brgemmOp = dyn_cast<xsmm::BrgemmOp>(oper);
+      break;
+    }
+  if (brgemmOp == NULL)
+    return failure();
```

Comment:
Is it safe to use when loop body is more complex?
Just wondering if there is anything about loop usage to be validated before we start modifying it.

Reply:
We check the shapes for tile shape mismatches as the next step, the op we look for is the innermost scf forall. We can probably add a few more checks like subview inputs for pytorch like models instead of mlir-gen like simple models, which adds to the complexity of our verifiers.

Reply:
Checks have been added for const buffers as well, as in the case of benchmarks.